### PR TITLE
Complete tracking and expiry of valid JWKs via the AuthenticatorState object

### PR DESCRIPTION
**Pull request by mystenmark**
I apologize for the large PR (although a lot of the + lines are docs and tests) - this turned out to be much more complicated than I could ever have guessed.

The main challenge here was managing the creation of the new system object at `0x7` that is used to track the authenticator state. It can't be created until one epoch after the protocol upgrade that brings in the new framework, and we have to be careful not to create it twice. I tried many different approaches before arriving at this one.

The system works as follows:
- `ChangeEpoch` is replaced with `EndOfEpochTransaction`, which contains a list of transactions to perform at the end of the epoch. For now the possible end-of-epoch actions are ChangeEpoch, AuthenticatorStateCreate, and AuthenticatorStateExpire.
- If the protocol feature flag is off, do nothing.
- If the feature flag is on, check if the authenticator state object already exists (and cache its existence for the duration of the epoch).
  - If it does not exist, then at the end of the epoch, add an `AuthenticatorStateCreate` tx just before the `ChangeEpoch` (inside EndOfEpochTransaction).
    - This was necessary because doing the creation from within `ChangeEpoch` posed the following difficulties:
      - ~~The start version of the authenticator state object would be variable instead of being fixed at 1.~~ Using EndOfEpochTransactions re-introduces this issue, which required adding the correct intial_shared_version to the transaction data.
      - I could not find an easy way to ensure that the creation could only happen once in this scenario.
    - This had the added benefit of giving a simple way to probe for existence of the `create` function. If the `AuthenticatorStateCreate` tx fails due to `FunctionNotFound` when attempting to augment the checkpoint, we simply omit the tx.
  - If it does exist, then we add an `AuthenticatorStateExpire` tx just before the `ChangeEpoch` tx.
    - I had similarly hoped to do the expiry from within the `ChangeEpoch` tx, but that would have required the set of shared_input_objects to the `ChangeEpoch` tx to change at a certain point. And because that point is delayed until the epoch after the protocol version bump (due to framework upgrade) it would have been extremely ugly to do this.
    - There would be no easy way to pass in a `min_epoch` argument to ChangeEpoch tx without add a ChangeEpochV2 transaction type with an additional argument. (as well as a new entry point to the `advance_epoch` function).
  - CheckpointExecutor has been modified to treat both of these "end of epoch" txns the way we previously treat `ChangeEpoch` - i.e. only execute it via CheckpointExecutor after all other transactions have been executed.
    - I tried to avoid this by having `handle_consensus_output` insert the `AuthenticatorStateExpire` tx, but this proved to be very awkward. It's not possible to tell whether the tx should be added until the `EndOfEpoch` messages have been processed, and by that point it is too late to add new system transactions.

The state machine for voting on JWKs, approving them, and created `AuthenticatorStateUpdate` transactions remains basically the same as in the previous version of this PR.

TODO: I am still working on a few tests, in particular one that provides multiple different JWKs from multiple providers to the TestCluster.

### CODE REVIEW

File: crates/sui-types/src/execution_mode.rs
Comment (lxfind) [inline crates/sui-types/src/execution_mode.rs:150]: Could you explain (and add comment) why we need this?

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (lxfind) [inline crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:64]: Why does this need to be public?

File: crates/sui-framework/packages/sui-system/sources/sui_system.move
Comment (lxfind) [inline crates/sui-framework/packages/sui-system/sources/sui_system.move:569]: We are shipping this with a different schedule to mainnet and testnet. But this will be always created at a particular protocol version?
An alternative way of doing this, which may be cleaner, is that in the Rust code when generating epoch change transaction, we could check whether the this feature is being enabled and if so, we insert a PT command into the epoch change transaction that creates the authenticator state.

File: crates/sui-core/src/authority.rs
Comment (lxfind) [inline crates/sui-core/src/authority.rs:2435]: Not a huge deal but could be safer to take the epoch store as argument here in this function.

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (mystenmark) [reply crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:64]: So that it can be called from sui_system - I tried to put this module in sui_system instead of sui and ran into some issues - i could try again.

File: crates/sui-framework/packages/sui-system/sources/sui_system.move
Comment (mystenmark) [reply crates/sui-framework/packages/sui-system/sources/sui_system.move:569]: I looked at doing that and it was very awkward to get the previous protocol config, but I will take another look. I agree that we should probably do that since as you point out, doing this without flags is a bit problematic

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:64]: I don't think this can safely be public. Then anyone will be able to create the object

Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:64]: Happy to help with whatever issues you ran into if it was in `sui_system` (which feels like a good home for this)

Comment (mystenmark) [reply crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:64]: @tnowacki it asserts that the sender is 0x0, so it should be safe. but i was able to remove the public anyway

File: crates/sui-core/src/authority.rs
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:2406]: this function is now removed

File: sui-execution/suivm/sui-adapter/src/execution_engine.rs
Comment (amnn) [inline sui-execution/suivm/sui-adapter/src/execution_engine.rs:528]: Heads up @dariorussi, when you merge `suivm` back into `latest`, you shouldn't merge this part.

Also, I'm guessing that (a) this change is going to land before your cut of `v1` and therefore you'll need to redo the cut.

File: crates/sui-rosetta/src/types.rs
Comment (amnn) [inline crates/sui-rosetta/src/types.rs:1]: @patrickkuo, what's the status of this crate -- is it still used, or planned to be used?

File: crates/sui-framework/packages/sui-system/sources/sui_system.move
Comment (amnn) [inline crates/sui-framework/packages/sui-system/sources/sui_system.move:81]: Is this used for anything? `const`s are private, so if it's not, then you may risk getting warnings about unused elements in the framework, in future.

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:48]: `String` means a UTF8 string, is that the right type here, vs an `std::ascii::String` or just a plain `vector<u8>`?

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:212]: How big can this list get?

File: crates/sui-types/src/execution_mode.rs
Comment (amnn) [reply crates/sui-types/src/execution_mode.rs:152]: I'm a bit conflicted on this point.  On the one hand, `allow_arbitrary_values` is quite a heavy hammer to wield for this, and you could get away with not needing it by flattening out `vector<ActiveJwk>`, a number of ways:

- A different vector for each field.
- A vector to pack all the string fields and then a separate vector for the epoch.
- BCS-encoded in a single `vector<u8>` (and then using `sui::bcs` to parse out the individual structs).

But on the other hand, that is understandably a pain.

**Review by amnn — COMMENTED** [review]
No blockers from me on the execution layer and move side.  I do hold out hope for a more general solution for adding new system objects, but I don't want to block this PR for that (in any case, the way the authenaticator state is updated may complicate things).

The only semi-substantive comment from me would be on maybe avoiding the need to allow arbitrary values for system transactions, but even that is not truly a blocker.

P.S. a completely left field question -- what would the solution look like if instead of getting its own ID, we added the authenticator state to the system state?

File: crates/sui-rosetta/src/types.rs
Comment (patrickkuo) [reply crates/sui-rosetta/src/types.rs:1]: this is currently being used by Coinbase

File: crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs
Comment (williampsmith) [inline crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs:579]: Could we add the digest to the log line here given that there are now more than one?

File: crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs
Comment (williampsmith) [inline crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs:871]: Does this need to be executed before the actual end of epoch txn? It looks to me like right now, given that this is pushed to the end for the vector, and then you call `execute_end_of_epoch_txn` twice serially, we will first enqueue  the actual end of epoch txn for execution, await its effects, then do the same for the authenticator state expiry txn.

If the two txns have a causal dependency, then it seems like you would end up deadlocking since they're in the same checkpoint and you enqueue the child and wait (and never get a chance to enqueue the parent).

If they're not, then don't we need to enforce that with the ordering here?

If order doesn't matter, then would it make sense to convert `execute_end_of_epoch_tx` to take a list, given that `handle_execution_effects` takes a list` so we could enqueue and await both in one shot

File: crates/sui-core/src/authority.rs
Comment (lxfind) [inline crates/sui-core/src/authority.rs:3761]: I think this is actually problematic given how authenticator_state_exists is cached in the epoch store. Since it's cached in the constructor, instead of truly being the state at the beginning of epoch, it not crash-safe.
You can have the object created and committed (when checkpoint executor executed this creation tx), then the node crashes, and when it wakes up it would think the object exists, then the validator path started to process the last checkpoint and it would fork.
The proper fix is likely to cache it in epoch start config.

Comment (mystenmark) [reply crates/sui-core/src/authority.rs:3761]: I'm confused - this was exactly the case I'm trying to prevent here. We don't re-check the object existence, we read the cached value from the db.

File: crates/sui-config/src/genesis.rs
Comment (andll) [inline crates/sui-config/src/genesis.rs:323]: nit: I would prefer `get_authenticator_state` to return Result<Option< AuthenticatorStateInner>> instead of converting any error to `Option::None`

File: crates/sui-core/src/authority.rs
Comment (andll) [reply crates/sui-core/src/authority.rs:3761]: I read this a bit more, and I think this should work - we snapshot whether object exists on the epoch store creation(so the function name is in fact confusing, it checks not that object exists, it checks that it existed in the beginning of the epoch), so it is essentially same as having a field in the EpochStartConfiguration.

I agree with the other comment from @lxfind that it probably would be cleaner to just put the field there, instead of creating new table though.

File: crates/sui-core/src/authority.rs
Comment (andll) [inline crates/sui-core/src/authority.rs:3783]: It seems that if this function fails, we will fail checkpoint creation as well.

This could be fine, if the only source of transaction being executed could be state sync(e.g. it should be safe to abort checkpoint builder at this point).

But, I think this can happen also simply in the failover scenario - imagine this function executed and node crashed. At this point transaction is executed, checkpoint builder will re-enter this path and fail creating the checkpoint...


I am not sure if this is actually important (e.g. do we really care about crash resilience in once-a-lifetime scenario), but wanted to point out anyway

**Comment by andll** [comment]
> If it does exist, then we add an AuthenticatorStateExpire tx just before the ChangeEpoch tx.
I had similarly hoped to do the expiry from within the ChangeEpoch tx, but that would have required the set of shared_input_objects to the ChangeEpoch tx to change at a certain point. And because that point is delayed until the epoch after the protocol version bump (due to framework upgrade) it would have been extremely ugly to do this.
There would be no easy way to pass in a min_epoch argument to ChangeEpoch tx without add a ChangeEpochV2 transaction type with an additional argument. (as well as a new entry point to the advance_epoch function).

So looks like potential alternative to second transaction is just having new ChangeEpochV2 with extra argument and migrate to using it after jwk state object has been created? If we had time I think I would chose this route, but I understand if we might not have time to make this change at this point

File: crates/sui-framework/packages/sui-system/sources/sui_system.move
Comment (mystenmark) [reply crates/sui-framework/packages/sui-system/sources/sui_system.move:81]: no, this is a left over from a previous PR, deleting

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (mystenmark) [reply crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:48]: I'm pretty sure JWKs can contain utf8 as they are just json

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (mystenmark) [reply crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:212]: Each validator can submit at most 2400 jwk votes per epoch, and we maintain the current + previous epoch's JWKs, so up to 4800. It is highly unlikely we would have this many. I may reduce the limit in fact, i think 2400 is probably too much.

File: crates/sui-core/src/authority.rs
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:3761]: Yes, I got confused and thought that EpochStartConfig was not versioned, and so using that would be difficult. But it is versioned, so i'm making that change right now.

File: crates/sui-core/src/authority.rs
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:3783]: This function never commits execution results to the DB. It only returns the effects so that we can construct the checkpoint. The commit happens only from checkpoint executor. So this should be safe. (And we also handle the case where the tx has already been executed via checkpoint at the point where we enter this function)

File: crates/sui-types/src/execution_mode.rs
Comment (mystenmark) [reply crates/sui-types/src/execution_mode.rs:152]: @amnn My inclination right now is to leave this as is, unless you see a strong argument against not allowing system txns to do this. When I spoke to @tnowacki about this earlier his take was "well, we disabled it earlier because we didn't need it, but there's no big reason not to allow it"

File: crates/sui-config/src/genesis.rs
Comment (mystenmark) [reply crates/sui-config/src/genesis.rs]: done!

File: crates/sui-core/src/authority.rs
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:3761]: done

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (mystenmark) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs]: done

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (amnn) [reply crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:212]: Yes, you probably don't want this getting much more than ~2000 for that total?

File: crates/sui-types/src/execution_mode.rs
Comment (amnn) [reply crates/sui-types/src/execution_mode.rs:163]: My main argument is that it means we need to do more due diligence around inputs to system transactions, because we could e.g. construct a coin and put it in there, breaking supply invariants.  We have checks and balances to catch supply invariants in particular, but generally it weakens the principle that all Move value constructors are protected within their modules.  To date we've only weakened that for dev-inspect which we don't persist results from.  Allowing arbitrary values in a transaction whose effects we do commit to the store felt like a different kettle of fish.

But, I may also be underestimating how special system transactions are already -- maybe this extra level of care we need to take is not significant relative to the other ways we need to treat system transactions specially.  If that's the case, feel free to ignore this comment.

**Comment by mystenmark** [comment]
> completely left field question -- what would the solution look like if instead of getting its own ID, we added the authenticator state to the system

@amnn It would be simpler in many ways, but I felt it was fairly important to avoid contention and dependencies on the system state object. Consider someone submitting a staking transaction - if their gas coin previous_transaction was missing, it could result in long delay for updating the authenticator state, during which time valid transactions might be rejected.

**Comment by mystenmark** [comment]
> So looks like potential alternative to second transaction is just having new ChangeEpochV2 with extra argument and migrate to using it after jwk state object has been created? If we had time I think I would chose this route, but I understand if we might not have time to make this change at this point

I switched to using `EndOfEpochTransaction` to encapsulate multiple end-of-epoch actions in a single transaction. Much simpler over all.

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (lxfind) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:122]: Might be worth adding a helper function in TransactionKind for this.

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (lxfind) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:564]: I thought we need to be able to tolerate the error here? Would the `?` cause early return?

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (lxfind) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:576]: Could we tolerate the error here?

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (lxfind) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:592]: ditto

File: crates/sui-core/src/authority.rs
Comment (lxfind) [inline crates/sui-core/src/authority.rs:2402]: Is this function used anywhere?

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (mystenmark) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:564]: you are right - this would have shown up as soon as I tested an upgrade on a local cluster. fixing

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (mystenmark) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:592]: yes

File: crates/sui-core/src/authority.rs
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:2402]: no, was from an earlier version of the PR, thanks for the catch

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (mystenmark) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:613]: For now, I don't think so, per our in-person discussion. I have pretty decent tests for the expire function so I think the risk of hitting this is low. I will investigate whether we can safely let it fail despite the shared temporary store. Added a TODO.

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:93]:
```suggestion
        (&a.jwk.kty == &b.jwk.kty) &&
           (&a.jwk.e == &b.jwk.e) &&
           (&a.jwk.n == &b.jwk.n) &&
           (&a.jwk.alg == &b.jwk.alg) &&
           (&a.jwk_id.iss == &b.jwk_id.iss) &&
           (&a.jwk_id.kid == &b.jwk_id.kid)
```

File: crates/sui-framework/packages/sui-system/sources/sui_system.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-system/sources/sui_system.move:54]: I don't see this change used anywhere? Did we not emit a warning?

File: crates/sui-framework/packages/sui-system/sources/sui_system.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-system/sources/sui_system.move:558]: nit spurious change makes source verification potentially sad (though not an issue since you are making other changes, just noting for the future)

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:571]: This makes me very uneasy, as we are creating/destroying a programmable transaction runtime multiple times. Which could lead to strange errors/interactions potentially in the future. In some ways, it feels like we will end up creating yet another execution loop, which is what programmable transactions was trying to prevent! 

I think instead, We should translate the `Vec<EndOfEpochTransactionKind>` to a single `ProgrammableTransaction` and then execute it.

**Review by tnowacki — DISMISSED** [review]
Requesting changes for the execution engine code. If there is some reason we cannot do a single PT, let me know and then dismiss this review :)

**Comment by mystenmark** [comment]
> Requesting changes for the execution engine code. If there is some reason we cannot do a single PT, let me know and then dismiss this review :)

The main reason is that we want to be able to tolerate failure of the AuthenticatorStateCreate block. However as I've been testing this today I'm reconsidering whether that is actually necessary. Will get back to you on that one.

File: crates/sui-framework/packages/sui-system/sources/sui_system.move
Comment (mystenmark) [reply crates/sui-framework/packages/sui-system/sources/sui_system.move:54]: I did not see one

File: crates/sui-framework/packages/sui-system/sources/sui_system.move
Comment (mystenmark) [reply crates/sui-framework/packages/sui-system/sources/sui_system.move:558]: my bad, reverting

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (mystenmark) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:573]: done

File: crates/sui-types/src/messages_consensus.rs
Comment (lxfind) [inline crates/sui-types/src/messages_consensus.rs:36]: Now that we do these in end of epoch, these types no longer belong to messages_consensus. Also we no longer need the epoch here.

File: crates/sui-types/src/messages_consensus.rs
Comment (lxfind) [inline crates/sui-types/src/messages_consensus.rs:44]: ditto

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (lxfind) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:329]: Does JWK become part of the key? That's quite non-intuitive. I don't have the context, but would be good to see a doc comment explanation at least.

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (lxfind) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:954]: What is this?? How does it pass tests?

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (lxfind) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:990]: ditto

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (mystenmark) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:329]: Yes, it's just storing a set of jwks with the round in which they became active, and its stored this way to make it easy to iterate by round.

The other option would be to have `DBMap<((u64, u64 /* index */), (JwkId, (JWK))>`, but this would require additional code to generate unique indices for each jwk.

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (mystenmark) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:954]: Leftover from before the EndOfEpochTransaction refactor, thanks for the catch. It passes because as far as I know, initial_shared_version is not used by the runtime, so you can pass in anything you want here. Fixing.

File: crates/sui-types/src/messages_consensus.rs
Comment (mystenmark) [reply crates/sui-types/src/messages_consensus.rs:36]: We do need the epoch, otherwise we'll have duplicated tx digests. (there is a comment above explaining this)

Moving to transactions.rs

File: crates/sui-types/src/messages_consensus.rs
Comment (mystenmark) [reply crates/sui-types/src/messages_consensus.rs]: needed for unique digests

File: crates/sui-types/src/messages_consensus.rs
Comment (lxfind) [reply crates/sui-types/src/messages_consensus.rs]: This is part of the EndOfEpoch tx now, which is guaranteed to have a unique tx digest since epoch is already part of it.

**Pull request merged**
